### PR TITLE
Add virtual fund sequence diagram

### LINKED ISSUE #108: commit a sequence diagram for the virtual-fund protocol

### PR DESCRIPTION
**Pull request by geoknee**
Closes #108 

TODO
- [x]  add information about cutting the number of rounds down (we know how to do that, it comes at some nonzero costs in terms of complexity).

### CODE REVIEW

File: protocols/virtual-fund/readme.md [hunk 1]
```diff
@@ -0,0 +1,17 @@
+# Virtual fund off-chain protocol
+
+## Single hop case
+
+Take three actors Alice, Bob and Irene. Given a ledger channel `L` between Alice and Irene and a ledger channel `L'` between Bob and Irene, the clients send and wait on messages as shown in the below sequence diagram in order to create and fund a virtual channel `V`:
```
Comment (andrewgordstewart) [inline protocols/virtual-fund/readme.md:5]: I tend to use `L` and `L'` due to laziness and my math background, but I think it's less common for programmers and maybe it's less clear for permanent documentation.

`L_a` and `L_b` might be better names. What do you think?

File: protocols/virtual-fund/readme.md [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+![](./virtual-fund-sequence-diagram.svg)
+
+The diagram is generated at https://sequencediagram.org/. The source code for this diagram is co-located in this folder.
```
Comment (andrewgordstewart) [inline protocols/virtual-fund/readme.md:9]: This addendum might be worth it.

```suggestion
The diagram is generated at https://sequencediagram.org/. The source code for this diagram is co-located in this folder, and should be updated in concert with changing the diagram.
```

**Review by andrewgordstewart — APPROVED** [review]
I think it's fine to merge this in as is, but do you have an opinion on whether we should be specializing to "Bob's initial balance is 0"?

File: protocols/virtual-fund/readme.md [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+## Number of rounds
+
+The number of rounds shown in this diagram is currently 4. The number of rounds can be reduced via a variety of optimizations, some of which are state below:
+
+- Round 0 and round 1 can be combined into a single round if all parties already know they want to join the virtual channel (rather than one party proposing it to the others).
+
+- Round 3 can be dispensed with entirely if Bob's initial balance is zero.
```
[the remark this answers is not in the document]
Comment (geoknee) [reply protocols/virtual-fund/readme.md:17]: I think we should implement the specialized protocol and make a corresponding diagram. The beauty of our architecture is that we can give that protocol a name, and have it live in the codebase alongside the existing code (and diagram). We can then make the switch between those protocols easily, or even have them both running in production (in principle). 

Once we have that implementation, the difference in code simplicity should be very clear to see. 

What do you think?

File: protocols/virtual-fund/readme.md (hunk 1 above)
Comment (geoknee) [reply protocols/virtual-fund/readme.md:5]: Sure, seems like an easy fix.

File: protocols/virtual-fund/virtual-fund.sequencediagram [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+parallel
+note left of A : <background:#yellow>Irene votes to fund V with L
+I->A: 2a: L.fundV(I)
+note right of B : <background:#yellow>Irene votes to fund V with L'
+I->B: 2b: L'.fundV(I)
+parallel off
+
+parallel 
+note left of A : <background:#yellow>Alice votes to fund V with L
```
[the remark this answers is not in the document]
Comment (geoknee) [reply protocols/virtual-fund/virtual-fund.sequencediagram:36]: I think it has its usual meaning here?

File: protocols/virtual-fund/readme.md (hunk 1 above)
Comment (geoknee) [reply protocols/virtual-fund/readme.md:5]: On second thought -- L_a is kind of ugly when not rendered in LaTeX. I minded to keep it as is until we hear from someone who actively prefers something else?

**Pull request merged**